ospfd/ospf_vty.c

```diff
@@ -9989,81 +9988,59 @@ DEFPY(ospf_gr_helper_planned_only,
 }
 
 /* External Route Aggregation */
-DEFUN (ospf_external_route_aggregation,
-       ospf_external_route_aggregation_cmd,
-       "summary-address A.B.C.D/M [tag (1-4294967295)]",
-       "External summary address\n"
-       "Summary address prefix\n"
-       "Router tag \n"
-       "Router tag value\n")
+DEFPY(ospf_external_route_aggregation,
+      ospf_external_route_aggregation_cmd,
+      "[no] summary-address A.B.C.D/M$prefix [{tag (1-4294967295)$tag_val | metric (0-16777215)$m_val | metric-type (1-2)$m_type}]",
+      NO_STR
+      "External summary address\n"
+      "Summary address prefix (a.b.c.d/m) \n"
+      "Router tag \n"
+      "Router tag value\n"
+      "Metric \n"
+      "Advertised metric for this route\n"
+      "OSPF exterior metric type for summarised routes\n"
+      "Set OSPF External Type 1/2 metrics\n")
 {
 	VTY_DECLVAR_INSTANCE_CONTEXT(ospf, ospf);
-	struct prefix_ipv4 p;
-	int idx = 1;
-	route_tag_t tag = 0;
+	struct prefix p;
 	int ret = OSPF_SUCCESS;
 
-	str2prefix_ipv4(argv[idx]->arg, &p);
-
-	if (is_default_prefix4(&p)) {
-		vty_out(vty,
-			"Default address shouldn't be configured as summary address.\n");
-		return CMD_SUCCESS;
-	}
-
-	/* Apply mask for given prefix. */
-	apply_mask((struct prefix *)&p);
-
-	if (!is_valid_summary_addr(&p)) {
-		vty_out(vty, "Not a valid summary address.\n");
+	ret = str2prefix(prefix_str, &p);
+	if (!ret) {
+		vty_out(vty, "%% Malformed prefix\n");
 		return CMD_WARNING_CONFIG_FAILED;
 	}
 
-	if (argc > 2)
-		tag = strtoul(argv[idx + 2]->arg, NULL, 10);
-
-	ret = ospf_asbr_external_aggregator_set(ospf, &p, tag);
-	if (ret == OSPF_INVALID)
-		vty_out(vty, "Invalid configuration!!\n");
-
-	return CMD_SUCCESS;
-}
-
-DEFUN (no_ospf_external_route_aggregation,
-       no_ospf_external_route_aggregation_cmd,
-       "no summary-address A.B.C.D/M [tag (1-4294967295)]",
-       NO_STR
-       "External summary address\n"
-       "Summary address prefix\n"
-       "Router tag\n"
-       "Router tag value\n")
-{
-	VTY_DECLVAR_INSTANCE_CONTEXT(ospf, ospf);
-	struct prefix_ipv4 p;
-	int idx = 2;
-	route_tag_t tag = 0;
-	int ret = OSPF_SUCCESS;
-
-	str2prefix_ipv4(argv[idx]->arg, &p);
-
-	if (is_default_prefix4(&p)) {
+	if (is_default_prefix4((struct prefix_ipv4 *)&p)) {
 		vty_out(vty,
 			"Default address shouldn't be configured as summary address.\n");
 		return CMD_SUCCESS;
 	}
 
 	/* Apply mask for given prefix. */
-	apply_mask((struct prefix *)&p);
+	apply_mask(&p);
 
-	if (!is_valid_summary_addr(&p)) {
+	if (!is_valid_summary_addr((struct prefix_ipv4 *)&p)) {
 		vty_out(vty, "Not a valid summary address.\n");
 		return CMD_WARNING_CONFIG_FAILED;
 	}
 
-	if (argc > 3)
-		tag = strtoul(argv[idx + 2]->arg, NULL, 10);
+	if (!m_val)
```

Comment:
m_val == 0 is allowed in CLI. This check should be replaced by `if (!m_val_str)`

Reply:
done